#ifndef MANTIDQTCUSTOMINTERFACES_CONVERTTOENERGY_H_
#define MANTIDQTCUSTOMINTERFACES_CONVERTTOENERGY_H_

//----------------------
// Includes
//----------------------
#include "ui_IndirectDataReduction.h"
#include "MantidQtAPI/AlgorithmRunner.h"
#include "MantidQtAPI/UserSubWindow.h"


namespace MantidQt
{
  namespace CustomInterfaces
  {
    //-------------------------------------------
    // Forward declarations
    //-------------------------------------------
    class Indirect;

    /** 
    This class defines the IndirectDataReduction interface. It handles the overall instrument settings
    and sets up the appropriate interface depending on the deltaE mode of the instrument. The deltaE
    mode is defined in the instrument definition file using the "deltaE-mode".    

    @author Martyn Gigg, Tessella Support Services plc
    @author Michael Whitty

    Copyright &copy; 2010 ISIS Rutherford Appleton Laboratory & NScD Oak Ridge National Laboratory

    This file is part of Mantid.

    Mantid is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    Mantid is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

    File change history is stored at: <https://github.com/mantidproject/mantid>
    Code Documentation is available at: <http://doxygen.mantidproject.org>    
    */

    class IndirectDataReduction : public MantidQt::API::UserSubWindow
    {
      Q_OBJECT

    public: // public constructor, destructor and functions
      /// Default Constructor
      IndirectDataReduction(QWidget *parent = 0);
      ///Destructor
      ~IndirectDataReduction();
      /// Interface name
      static std::string name() { return "Indirect Data Reduction"; }
      // This interface's categories.
      static QString categoryInfo() { return "Indirect"; }

    private slots:
      void helpClicked();
      void runClicked();
      void instrumentLoadingDone(bool error);
      void instrumentLoadProgress(double p, const std::string &msg);

    private: // private functions (and slots)
      /// Initialize the layout
      virtual void initLayout();
      virtual void initLocalPython();
      void readSettings();
      void saveSettings();
      void setDefaultInstrument(const QString & name);
      void instrumentSelectChanged(const QString& name);
      ///// Find path to instrument's _Definition.xml file (and check there is a parameter file).
      //QString getIDFPath(const QString& prefix);

    private slots:
      void userSelectInstrument(const QString& name);
      void openDirectoryDialog();

    private: // member variables
      /// The .ui form generated by Qt Designer
      Ui::IndirectDataReduction m_uiForm;
      /// Indirect Instruments interface object
      Indirect *m_indirectInstruments;
      /// Instrument the interface is currently set for.
      QString m_curInterfaceSetup;
      /// The settings group
      QString m_settingsGroup;
      /// Runner for insturment load algorithm
      MantidQt::API::AlgorithmRunner* m_algRunner;
    };

  }
}

#endif //MANTIDQTCUSTOMINTERFACES_CONVERTTOENERGY_H_
